python/ray/tune/schedulers/pb2.py

```diff
@@ -0,0 +1,341 @@
+        current = np.hstack((padding, current))
+
+        Xnew = np.vstack((X, current))
+        ypad = np.zeros(current.shape[0])
+        ypad = ypad.reshape(-1, 1)
+        ynew = np.vstack((y, ypad))
+
+        #kernel = GPy.kern.RBF(input_dim=X.shape[1], variance=1., lengthscale=1.)
+        kernel = TV_SquaredExp(
+            input_dim=X.shape[1], variance=1., lengthscale=1., epsilon=0.1)
+        m1 = GPy.models.GPRegression(Xnew, ynew, kernel)
+        m1.optimize()
+
+    xt = optimize_acq(UCB, m, m1, fixed, num_f)
+
+    # convert back...
+    xt = xt * (np.max(base_vals, axis=0) - np.min(base_vals, axis=0)) + np.min(
+        base_vals, axis=0)
+
+    xt = xt.astype(np.float32)
+    return (xt)
+
+
+def explore(data, bounds, current, base, old, config):
+    """ This function primarily processes the data from completed trials
+        and then requests the next config from the select_config function.
+        It then adds the new trial to the dataframe, so that the reward change
+        can be computed using the new weights.
+        It returns the new point and the dataframe with the new entry.
+    """
+
+    df = data.sort_values(by='Time').reset_index(drop=True)
+
+    # group by trial ID and hyperparams. Compute change in timesteps and reward.
+    df['y'] = df.groupby(['Trial'] +
+                         [key for key in bounds.keys()])['Reward'].diff()
+    df['t_change'] = df.groupby(['Trial'] +
+                                [key for key in bounds.keys()])['Time'].diff()
+
+    # delete entries without positive change in t.
+    df = df[df['t_change'] > 0].reset_index(drop=True)
+    df['R_before'] = df.Reward - df.y
+
+    # normalize the reward change by the update size. For example if trials took diff lengths of time.
+    df['y'] = df.y / df.t_change
+    df = df[~df.y.isna()].reset_index(drop=True)
+    df = df.sort_values(by='Time').reset_index(drop=True)
+
+    # only use the last 1k datapoints, so the GP is not too slow
+    df = df.iloc[-1000:, :].reset_index(drop=True)
+
+    # we need this to know the T and Reward for the weights
+    dfnewpoint = df[df['Trial'] == str(base)]
+
+    if not dfnewpoint.empty:
+
+        # now specify the dataset for the GP
+        y = np.array(df.y.values)
+        # meta data we keep -> episodes and reward (TODO: convert to curve)
+        t_r = df[['Time', 'R_before']]  
+        h = df[[key for key in bounds.keys()]]
+        X = pd.concat([t_r, h], axis=1).values
+        newpoint = df[df['Trial'] == str(base)].iloc[-1, :][['Time', 'R_before'
+                                                             ]].values
+        new = select_config(
+            X,
+            y,
+            current,
+            newpoint,
+            bounds,
+            num_f=len(t_r.columns),
+            num=len(h.columns))
+
+        new_config = config.copy()
+        values = []
+        names = []
+        for i, col in enumerate(h.columns):
+            if type(config[col]) is int:
```

Comment:
should we do this? probably won't make much of a difference
```suggestion
            if isinstance(config[col], int):
```